x/examples/test-connectivity/main.go

```diff
@@ -161,56 +211,141 @@ func main() {
 	success := false
 	jsonEncoder := json.NewEncoder(os.Stdout)
 	jsonEncoder.SetEscapeHTML(false)
-	configToDialer := configurl.NewDefaultConfigToDialer()
 	for _, resolverHost := range strings.Split(*resolverFlag, ",") {
 		resolverHost := strings.TrimSpace(resolverHost)
 		resolverAddress := net.JoinHostPort(resolverHost, "53")
 		for _, proto := range strings.Split(*protoFlag, ",") {
 			proto = strings.TrimSpace(proto)
 			var resolver dns.Resolver
+			var mu sync.Mutex
+			dnsReports := make([]dnsReport, 0)
+			tcpReports := make([]tcpReport, 0)
+			configToDialer := configurl.NewDefaultConfigToDialer()
+			configToDialer.BaseStreamDialer = transport.FuncStreamDialer(func(ctx context.Context, addr string) (transport.StreamConn, error) {
+				hostname, _, err := net.SplitHostPort(addr)
+				if err != nil {
+					return nil, err
+				}
+				onDNS := func(ctx context.Context, domain string) func(di httptrace.DNSDoneInfo) {
+					dnsStart := time.Now()
+					return func(di httptrace.DNSDoneInfo) {
+						report := dnsReport{
+							QueryName:  hostname,
+							Time:       dnsStart.UTC().Truncate(time.Second),
+							DurationMs: time.Since(dnsStart).Milliseconds(),
+						}
+						if di.Err != nil {
+							report.Error = di.Err.Error()
+						}
+						for _, ip := range di.Addrs {
+							report.AnswerIPs = append(report.AnswerIPs, ip.IP.String())
+						}
+						mu.Lock()
+						dnsReports = append(dnsReports, report)
+						mu.Unlock()
+					}
+				}
+				onDial := func(ctx context.Context, network, addr string, connErr error) {
+					ip, port, err := net.SplitHostPort(addr)
+					if err != nil {
+						return
+					}
+					report := tcpReport{
+						Hostname: hostname,
+						IP:       ip,
+						Port:     port,
+					}
+					if connErr != nil {
+						report.Error = connErr.Error()
+					}
+					mu.Lock()
+					tcpReports = append(tcpReports, report)
+					mu.Unlock()
+				}
+				return newTCPTraceDialer(onDNS, onDial).DialStream(ctx, addr)
+			})
+			configToDialer.BasePacketDialer = transport.FuncPacketDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+				hostname, _, err := net.SplitHostPort(addr)
+				if err != nil {
+					return nil, err
+				}
+				var dnsStart time.Time
+				ctx = httptrace.WithClientTrace(ctx, &httptrace.ClientTrace{
+					DNSStart: func(di httptrace.DNSStartInfo) {
+						dnsStart = time.Now()
+					},
+					DNSDone: func(di httptrace.DNSDoneInfo) {
+						report := dnsReport{
+							QueryName:  hostname,
+							Time:       dnsStart.UTC().Truncate(time.Second),
+							DurationMs: time.Since(dnsStart).Milliseconds(),
+						}
+						if di.Err != nil {
+							report.Error = di.Err.Error()
+						}
+						for _, ip := range di.Addrs {
+							report.AnswerIPs = append(report.AnswerIPs, ip.IP.String())
+						}
+						// TODO(fortuna): Use a Mutex.
```

Comment:
```suggestion
mu.Lock()
```

Reply:
Done in https://github.com/Jigsaw-Code/outline-sdk/pull/272/commits/acdc62201c16952ebbf4176759a2d8c9f33fa8a8